### PR TITLE
Create E1.31 light effect for non-addressable lights

### PR DESCRIPTION
**Pull request by Almighty-Alpaca**
# What does this implement/fix?

This PR implements the E1.31 light effect for non-addressable lights. This was previously only possible with a weird workaround by creating a partition with a length of 1.

To stick to the naming conventions of other light effects the addressable variant was renamed to `addressable_e131`, making this a breaking change. `e131` is now the non-addressable effect. 

Since I do not currently have an addressable RGB strip I was only able to test the `addressable_e131` effect with the partition workaround. The non-addressable effect has been in use for many months in my personal setup without any issues.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# To test this change without checking out the PR in Git
external_components:
  - source: "github://Almighty-Alpaca/esphome@simple-e131-effect"
    components: [ "e131" ]

e131:
  method: multicast

light:
  - platform: rgb
    name: "RGB Light"
    id: led
    red: led_r
    green: led_g
    blue: led_b
    effects:
      - e131:
          universe: 1
          channels: RGB
  - platform: neopixelbus
    variant: WS2811
    pin: GPIO23
    num_leds: 60
    name: "ARGB Light"
    effects:
      - addressable_e131:
          universe: 2
          channels: RGB

output:
  - platform: ledc
    id: led_g
    pin: GPIO13
  - platform: ledc
    id: led_r
    pin: GPIO12
  - platform: ledc
    id: led_b
    pin: GPIO14
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/e131/e131_light_effect_base.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  void set_channels(E131LightChannels channels) { this->channels_ = channels; }
+  void set_e131(E131Component *e131) { this->e131_ = e131; }
+
+ protected:
+  int first_universe_{0};
+  int last_universe_{0};
+  E131LightChannels channels_{E131_RGB};
+  E131Component *e131_{nullptr};
+
+  virtual bool process_(int universe, const E131Packet &packet) = 0; // NOLINT(readability-identifier-naming)
```
Comment (jesserockz) [inline esphome/components/e131/e131_light_effect_base.h]:
```suggestion
  virtual bool process(int universe, const E131Packet &packet) = 0;
```

File: esphome/components/e131/e131_light_effect.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  const std::string &get_name() override;
+
+  void start() override;
+  void stop() override;
+  void apply() override;
+
+  int get_universe_count() const override;
+
+ protected:
+  bool process_(int universe, const E131Packet &packet) override;
```
Comment (jesserockz) [inline esphome/components/e131/e131_light_effect.h]:
```suggestion
  bool process(int universe, const E131Packet &packet) override;
```

File: esphome/components/e131/e131_addressable_light_effect.h [hunk 3]
```diff
@@ -4,42 +4,26 @@
...
  protected:
-  bool process_(int universe, const E131Packet &packet);
-
-  int first_universe_{0};
-  int last_universe_{0};
-  E131LightChannels channels_{E131_RGB};
-  E131Component *e131_{nullptr};
-
-  friend class E131Component;
+  bool process_(int universe, const E131Packet &packet) override;
```
Comment (jesserockz) [inline esphome/components/e131/e131_addressable_light_effect.h]:
```suggestion
  bool process(int universe, const E131Packet &packet) override;
```

File: esphome/components/e131/e131_light_effect.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  LightEffect::stop();
+}
+
+void E131LightEffect::apply() {
+  // ignore, it is run by `E131Component::update()`
+}
+
+int esphome::e131::E131LightEffect::get_universe_count() const { return 1; }
+
+bool esphome::e131::E131LightEffect::process_(int universe, const E131Packet &packet) {
```
Comment (jesserockz) [inline esphome/components/e131/e131_light_effect.cpp]:
```suggestion
bool esphome::e131::E131LightEffect::process(int universe, const E131Packet &packet) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Virtual protected methods don't have trailing underscores

One more I could not add a review for on https://github.com/esphome/esphome/pull/3932/files#diff-78af0c8f4d80a0219497b177ac521e2edba7e350d92c96f6e59da29a769ef30aR36

**Comment by Almighty-Alpaca** [comment]
I've implemented the requested changes and also gone ahead and rebased the PR onto the latest commit on the dev branch.

**Review by jesserockz — CHANGES_REQUESTED** [review]
I don't have the capacity to investigate, but could you try to see if both addressable and non-addressable could be named `e131` and make a determination in codegen which class should be used? 

If the answer is "that won't work" then that is acceptable and I will have another think about changing `e131` to `addressable_e131`

**Pull request closed without merging**